arch/arm/src/nrf52/nrf52_sdc.c

```diff
@@ -0,0 +1,609 @@
+                       (SDC_MASTER_COUNT * MASTER_MEM_SIZE))
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+struct nrf52_sdc_dev_s
+{
+  uint8_t mempool[MEMPOOL_SIZE];
+  uint8_t msg_buffer[HCI_MSG_BUFFER_MAX_SIZE];
+
+  sem_t exclsem;
+  struct work_s work;
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static void mpsl_assert_handler(const char *const file, const uint32_t line);
+static void sdc_fault_handler(const char *file, const uint32_t line);
+
+static int bt_open(FAR const struct bt_driver_s *btdev);
+static int bt_hci_send(FAR const struct bt_driver_s *btdev,
+                       FAR struct bt_buf_s *buf);
+
+static void on_hci(void);
+static void on_hci_worker(void *arg);
+
+static void low_prio_worker(void *arg);
+
+static int swi_isr(int irq, FAR void *context, FAR void *arg);
+static int power_clock_isr(int irq, FAR void *context, FAR void *arg);
+
+static void rng_handler(void);
+static void rtc0_handler(void);
+static void timer0_handler(void);
+static void radio_handler(void);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static const struct bt_driver_s g_bt_driver =
+{
+  .head_reserve = 0,
+  .open         = bt_open,
+  .send         = bt_hci_send
+};
+
+static const mpsl_clock_lfclk_cfg_t g_clock_config =
+{
+  .source                   = MPSL_CLOCK_LF_SRC_XTAL,
+  .rc_ctiv                  = 0,
+  .rc_temp_ctiv             = 0,
+  .accuracy_ppm             = CONFIG_NRF52_SDC_CLOCK_ACCURACY,
+  .skip_wait_lfclk_started  = false
+};
+
+static struct nrf52_sdc_dev_s g_sdc_dev;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: bt_open
+ ****************************************************************************/
+
+static int bt_open(FAR const struct bt_driver_s *btdev)
+{
+  return 0;
+}
+
+/****************************************************************************
+ * Name: bt_open
+ ****************************************************************************/
+
+static int bt_hci_send(FAR const struct bt_driver_s *btdev,
+                       FAR struct bt_buf_s *buf)
+{
+  int ret = OK;
+  irqstate_t flags;
+
+  /* Pass HCI CMD/DATA to SDC */
+
+  if (buf->type == BT_CMD)
+    {
+      struct bt_hci_cmd_hdr_s *cmd = (struct bt_hci_cmd_hdr_s *)buf->data;
+
+      wlinfo("passing CMD %d to softdevice\n", cmd->opcode);
+
+      /* Ensure non-concurrent access to SDC operations */
```

Comment:
We should use a semaphore to control locking of SDC, this will lock the processor which we don't want to do.

Reply:
I see you already have g_sdc_dev.exclsem  why not use this?

Reply:
That is what I mentioned before I was unsure about. Please have a look at:
- https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrfxlib/mpsl/doc/mpsl.html#low-priority
- https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrfxlib/softdevice_controller/doc/softdevice_controller.html#integration-with-applications

My interpretation is that we need to protect these calls from being interrupted, not only avoid concurrent calls. What do you think?

Reply:
I'll take a look I think we can be more loose here.

Reply:
I'll take a look I think we can be more loose here.

Reply:
From looking these docs I really think that the only thing you need to do is use the semaphore to protect the calls to sdc/mpsl.  The only restriction is you cannot have API calls interrupting each other, but you can have the API calls be interrupted (this is explicitly allowed).  The note about disabling interrupts I read to be one way to meet the synchronisation requirements (although not true in SMP).   I think this fits squarely in the last note:
`Alternatively, you can use synchronization primitives to ensure that no MPSL functions are called at the same time.`

So simply dropping the critical section here would be totally reasonably to me.

Reply:
That whole paragraph is a bit confusing to me. Since the "alternatively" is not clear about what is referring to. One of the previous bullets states:

  > Application must only call MPSL APIs from non-preemptible threads, or with interrupts disabled (e.g. during initialization).

which is what confuses me. Anyway, if you look at NRF SDK code, there only seems to be a semaphore protecting the calls. So I guess that should be enough. I will do the change and do some more stress testing to see if something strange comes up.

Reply:
I agree its poorly worded, I think what they are trying to cover here is where an "application" rather than the OS is making calls to the API (say two user threads running a ble app that talks directly via the APIs)  in that case they likely would likely not know they need to share a synchronization primitive.  Here we have a OS driver so we have a lot more control and can be sure the APIs are locked.

Reply:
I removed the critical sections and it seems things remain stable. I pushed this change.